File: core/codec/decode.ts
// WO-1082-C canonical decoder — validation order mirrors codec/src/decode.rs exactly.
// Uses DataView for explicit LE reads. No implicit coercions. No float arithmetic.
// Terminal invariant: cursor === frame_length.

import { ABI_VERSION, DOMAIN_SEPARATOR, HEADER_SIZE, EVENT_HEADER_SIZE, MAX_BATCH_SIZE, MAX_FRAME_SIZE } from './constants.js';
import { CodecError, DecodeError } from './errors.js';
import { Event, Frame } from './types.js';

export function decode(buf: Uint8Array): Frame {
    const len  = buf.length;
    const view = new DataView(buf.buffer, buf.byteOffset, buf.byteLength);

    // 1. minimum frame length
    if (len < HEADER_SIZE) {
        throw new CodecError(DecodeError.InvalidLength);
    }

    // 2. maximum frame length
    if (len > MAX_FRAME_SIZE) {
        throw new CodecError(DecodeError.FrameSizeOutOfBounds);
    }

    // 3. frame_length equality — u32 LE at bytes [0..4]
    const frameLength = view.getUint32(0, true);
    if (frameLength !== len) {
        throw new CodecError(DecodeError.LengthMismatch);
    }

    // 4. version equality — byte [4]
    if (view.getUint8(4) !== ABI_VERSION) {
        throw new CodecError(DecodeError.UnsupportedVersion);
    }

    // 5. domain separator equality — byte [5]
    if (view.getUint8(5) !== DOMAIN_SEPARATOR) {
        throw new CodecError(DecodeError.DomainMismatch);
    }

    // frame domain_id — byte [6]
    const frameDomainId = view.getUint8(6);

    // event_count — u32 LE at bytes [7..11]
    const eventCount = view.getUint32(7, true);

    // 6. event count bounds
    if (eventCount > MAX_BATCH_SIZE) {
        throw new CodecError(DecodeError.ExcessiveEventCount);
    }

    // bytes [11..41] reserved — consumed, not validated in ABI v1.0

    let cursor = HEADER_SIZE;
    const events: Event[] = [];

    for (let i = 0; i < eventCount; i++) {
        // 7. 20-byte fixed event header boundary
        if (cursor + EVENT_HEADER_SIZE > frameLength) {
            throw new CodecError(DecodeError.TruncatedEventHeader);
        }

        // decode deterministic LE primitives
        const domainId    = view.getUint8(cursor);
        const typeId      = view.getUint8(cursor + 1);
        const commitIndex = view.getUint32(cursor + 2, true);
        const sequenceId  = view.getUint32(cursor + 6, true);
        const spatialKey  = view.getBigUint64(cursor + 10, true);
        const payloadLen  = view.getUint16(cursor + 18, true);

        cursor += EVENT_HEADER_SIZE;

        // validate domain_id — must be non-zero
        if (domainId === 0) {
            throw new CodecError(DecodeError.InvalidDomainId);
        }

        // validate payload bounds
        if (cursor + payloadLen > frameLength) {
            throw new CodecError(DecodeError.TruncatedPayload);
        }

        // copy payload — advance cursor strictly
        const payload = buf.slice(cursor, cursor + payloadLen);
        cursor += payloadLen;

        events.push({ domainId, typeId, commitIndex, sequenceId, spatialKey, payload });
    }

    // terminal invariant: cursor must equal frame_length exactly
    if (cursor !== frameLength) {
        throw new CodecError(DecodeError.LengthInvariantViolation);
    }

    return { domainId: frameDomainId, events };
}

File: codec/src/decode.rs
// WO-1082-B canonical decoder — header validation order is authoritative.
// Cursor traversal is strictly forward. Terminal invariant: cursor == frame_length.

use crate::constants::*;
use crate::errors::DecodeError;
use crate::types::{Event, Frame};

pub fn decode(buf: &[u8]) -> Result<Frame, DecodeError> {
    // 1. minimum frame length
    if buf.len() < HEADER_SIZE {
        return Err(DecodeError::InvalidLength);
    }

    // 2. maximum frame length
    if buf.len() > MAX_FRAME_SIZE {
        return Err(DecodeError::FrameSizeOutOfBounds);
    }

    // 3. frame_length equality — declared u32 LE at bytes [0..4]
    let frame_length = u32::from_le_bytes([buf[0], buf[1], buf[2], buf[3]]) as usize;
    if frame_length != buf.len() {
        return Err(DecodeError::LengthMismatch);
    }

    // 4. version equality — byte [4]
    if buf[4] != ABI_VERSION {
        return Err(DecodeError::UnsupportedVersion);
    }

    // 5. domain separator equality — byte [5]
    if buf[5] != DOMAIN_SEPARATOR {
        return Err(DecodeError::DomainMismatch);
    }

    // frame domain_id — byte [6]
    let frame_domain_id = buf[6];

    // event_count — u32 LE at bytes [7..11]
    let event_count = u32::from_le_bytes([buf[7], buf[8], buf[9], buf[10]]) as usize;

    // 6. event count bounds
    if event_count > MAX_BATCH_SIZE {
        return Err(DecodeError::ExcessiveEventCount);
    }

    // bytes [11..41] are reserved — consumed but not validated in ABI v1.0

    let mut cursor = HEADER_SIZE;
    let mut events = Vec::with_capacity(event_count);

    for _ in 0..event_count {
        // 7. structural divisibility / 20-byte fixed header boundary
        if cursor + EVENT_HEADER_SIZE > frame_length {
            return Err(DecodeError::TruncatedEventHeader);
        }

        // decode deterministic LE primitives
        let domain_id   = buf[cursor];
        let type_id     = buf[cursor + 1];
        let commit_index = u32::from_le_bytes([
            buf[cursor + 2], buf[cursor + 3], buf[cursor + 4], buf[cursor + 5],
        ]);
        let sequence_id = u32::from_le_bytes([
            buf[cursor + 6], buf[cursor + 7], buf[cursor + 8], buf[cursor + 9],
        ]);
        let spatial_key = u64::from_le_bytes([
            buf[cursor + 10], buf[cursor + 11], buf[cursor + 12], buf[cursor + 13],
            buf[cursor + 14], buf[cursor + 15], buf[cursor + 16], buf[cursor + 17],
        ]);
        let payload_len = u16::from_le_bytes([buf[cursor + 18], buf[cursor + 19]]) as usize;

        cursor += EVENT_HEADER_SIZE;

        // validate domain_id — must be non-zero
        if domain_id == 0 {
            return Err(DecodeError::InvalidDomainId);
        }

        // validate payload bounds
        if cursor + payload_len > frame_length {
            return Err(DecodeError::TruncatedPayload);
        }

        // copy payload — advance cursor strictly
        let payload = buf[cursor..cursor + payload_len].to_vec();
        cursor += payload_len;

        events.push(Event { domain_id, type_id, commit_index, sequence_id, spatial_key, payload });
    }

    // terminal invariant: cursor must be exactly frame_length
    if cursor != frame_length {
        return Err(DecodeError::LengthInvariantViolation);
    }

    Ok(Frame { domain_id: frame_domain_id, events })
}
